python/tskit/combinatorics.py

```diff
@@ -0,0 +1,723 @@
+        ]
+
+        t = RankTree(children=labelled_children)
+        t._shape_rank = self.shape_rank()
+        t._label_rank = label_rank
+        return t
+
+    @staticmethod
+    def canonical_order(c):
+        """
+        Defines the canonical ordering of sibling subtrees.
+        """
+        return c.num_leaves, c.shape_rank(), c.min_label()
+
+    @staticmethod
+    def from_tsk_tree_node(tree, u):
+        if tree.is_leaf(u):
+            return RankTree(children=[], label=u)
+
+        children = list(
+            sorted(
+                (RankTree.from_tsk_tree_node(tree, c) for c in tree.children(u)),
+                key=RankTree.canonical_order,
+            )
+        )
+        return RankTree(children=children)
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        if len(tree.roots) != 1:
+            raise ValueError("Can't rank trees with multiple roots")
+
+        return RankTree.from_tsk_tree_node(tree, tree.root)
+
+    def to_tsk_tree(self):
+        seq_length = 1
+        tables = tskit.TableCollection(seq_length)
+
+        def add_node(node):
+            if node.is_leaf():
+                assert node.label is not None
+                return node.label
+
+            child_ids = [add_node(child) for child in node.children]
+            # Arbitrarily set parent time +1 from their oldest child
+            max_child_time = max(tables.nodes.time[c] for c in child_ids)
+            parent_id = tables.nodes.add_row(time=max_child_time + 1)
+            for child_id in child_ids:
+                tables.edges.add_row(0, seq_length, parent_id, child_id)
+
+            return parent_id
+
+        for _ in range(self.num_leaves):
+            tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        add_node(self)
+
+        # Have to sort for now because the of the first way
+        # in which we're traversing the edges
+        tables.sort()
+        return tables.tree_sequence().first()
+
+    def newick(self):
+        if self.is_leaf():
+            return str(self.label) if self.labelled() else ""
+        return "(" + ",".join(c.newick() for c in self.children) + ")"
+
+    @property
+    def label(self):
+        return self.labels[0]
+
+    def labelled(self):
+        return all(l is not None for l in self.labels)
+
+    def min_label(self):
+        return self.labels[0]
+
+    def is_leaf(self):
+        return len(self.children) == 0
+
+    def leaf_partition(self):
+        return [c.num_leaves for c in self.children]
+
+    def group_children_by_num_leaves(self):
+        def same_num_leaves(c1, c2):
+            return c1.num_leaves == c2.num_leaves
+
+        return group_by(self.children, same_num_leaves)
+
+    def group_children_by_shape(self):
+        def same_shape(c1, c2):
+            return c1.num_leaves == c2.num_leaves and c1.shape_rank() == c2.shape_rank()
+
+        return group_by(self.children, same_shape)
+
+    def equal(self, other):
```

Comment:
Could consider overriding ``__eq__`` and ``__neq__`` here. Seems like a good definition of equality.